### PR TITLE
003 view job capabilities

### PR DESCRIPTION
**Pull request by dmkainos**

### CODE REVIEW

**Review by dcorry586 — APPROVED** [review]

File: sql-scripts/create-job-capabilities.sql [hunk 1]
```diff
@@ -0,0 +1,36 @@
+USE TheITCrowd_MiaG;
+
+CREATE TABLE JobCapabilities(
+    jobCapabilitiesId INT AUTO_INCREMENT,
```
Comment (glenn-hamilton) [inline sql-scripts/create-job-capabilities.sql:4]: Is `INT` the correct data type?

File: sql-scripts/create-job-capabilities.sql [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO JobCapabilities (jobCapabilitiesName) VALUES ('Delivery');
+INSERT INTO JobCapabilities (jobCapabilitiesName) VALUES ('Operations');
+INSERT INTO JobCapabilities (jobCapabilitiesName) VALUES ('Business Development and Marketing');
+INSERT INTO JobCapabilities (jobCapabilitiesName) VALUES ('Organisational Strategy and Planning');
+INSERT INTO JobCapabilities (jobCapabilitiesName) VALUES ('People');
+INSERT INTO JobCapabilities (jobCapabilitiesName) VALUES ('Commercial and Financial Management');
+INSERT INTO JobCapabilities (jobCapabilitiesName) VALUES ('Business Services Support');
+
+ALTER TABLE JobRoles
+ADD COLUMN jobCapabilityId INT;
```
Comment (glenn-hamilton) [inline sql-scripts/create-job-capabilities.sql:25]: Is `INT` the correct data type?

File: src/main/java/org/kainos/ea/cli/JobCapabilityRequest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  public String getJobCapabilityName() {
+    return jobCapabilityName;
+  }
+
+  public void setJobCapabilityName(String jobCapabilityName) {
+    this.jobCapabilityName = jobCapabilityName;
+  }
+
+
+
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/cli/JobCapabilityRequest.java]: Formatting

File: src/main/java/org/kainos/ea/controller/JobCapabilityController.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  @GET
+  @Path("/job-capabilities")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getJobCapabilities() {
+    try {
+      return Response.ok(jobCapabilityService.getAllJobCapabilities()).build();
+    } catch (SQLException e) {
+      System.err.println(e);
+      return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+              .entity("Database error: " + e.getMessage())
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/controller/JobCapabilityController.java]: You don't want to pass this back

File: src/main/java/org/kainos/ea/controller/JobCapabilityController.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+              .build();
+    } catch (CouldNotGetJobCapabilitiesException e) {
+      System.err.println(e);
+      return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+              .entity("Job Capabilities retrieval error: " + e.getMessage())
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/controller/JobCapabilityController.java]: Probably don't want to pass this back either

File: src/main/java/org/kainos/ea/controller/JobCapabilityController.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  @GET
+  @Path("/job-capabilities/{jobCapabilityId}")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getJobCapabilityById(@PathParam("jobCapabilityId") int jobCapabilityId) {
+    try {
+      return Response.status(HttpStatus.OK_200).entity(jobCapabilityService.getJobCapability(jobCapabilityId)).build();
+    } catch (SQLException e) {
+      System.err.println(e);
+      return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+              .entity("Database error: " + e.getMessage())
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/controller/JobCapabilityController.java]: As above

File: src/main/java/org/kainos/ea/controller/JobCapabilityController.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+              .build();
+    } catch (JobCapabilityDoesNotExist e) {
+      System.err.println(e);
+      return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+              .entity("Job Capability Error: " + e.getMessage())
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/controller/JobCapabilityController.java]: As above

File: src/main/java/org/kainos/ea/db/JobCapabilityDao.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        jobCapabilities.add(jobCapability);
+      }
+      return jobCapabilities;
+    }
+  }
+
+  public JobCapabilityRequest getJobCapability(int jobCapabilityId, Connection c) throws SQLException {
+    try (Statement st = c.createStatement()) {
+
+      ResultSet rs = st.executeQuery("SELECT jobCapabilityName FROM JobCapabilities WHERE jobCapabilityId =" + jobCapabilityId);
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/db/JobCapabilityDao.java]: SQLInjection risk. Use a prepared statement

File: src/main/java/org/kainos/ea/db/JobCapabilityDao.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+      while (rs.next()) {
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/db/JobCapabilityDao.java]: This doesn't return a list, can use an `if` instead o a `while`

**Review by dcorry586 — APPROVED** [review]

File: src/main/java/org/kainos/ea/controller/JobCapabilityController.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    }
+  }
+
+  @GET
+  @Path("/job-capabilities/{jobCapabilityId}")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getJobCapabilityById(@PathParam("jobCapabilityId") int jobCapabilityId) {
+    try {
+      return Response.status(HttpStatus.OK_200).entity(jobCapabilityService.getJobCapability(jobCapabilityId)).build();
+    } catch (SQLException | JobCapabilityDoesNotExist e) {
```
Comment (glenn-hamilton) [inline src/main/java/org/kainos/ea/controller/JobCapabilityController.java]: Should `JobCapabilityDoesNotExist` not `404`?

**Review by glenn-hamilton — APPROVED** [review]

**Pull request merged**